timemachine/maps/terminal_bonds.py

```diff
@@ -0,0 +1,236 @@
+
+@jit
+def interval_map(x, src_lb, src_ub, dst_lb, dst_ub):
+    """map x in Interval[src_lb, src_ub] to y in Interval[dst_lb, dst_ub]"""
+
+    scale_factor = (dst_ub - dst_lb) / (src_ub - src_lb)
+
+    in_support = (x >= src_lb) & (x <= src_ub)
+    mapped = dst_lb + (x - src_lb) * scale_factor
+
+    return jnp.where(in_support, mapped, np.nan)
+
+
+def conf_map(x, bond, param) -> Tuple[Array, Array]:
+    """Apply map to a single bond in a conformer x, return (updated x, logdetjac)"""
+    a, b = bond
+    dim = 3
+
+    def f(xy, param):
+        """R^[2 x dim] -> R^[2 x dim]"""
+        x, y = xy[:dim], xy[dim:]
+        src_lb, src_ub, dst_lb, dst_ub = param
+
+        r = jnp.linalg.norm(x - y)
+        new_r = interval_map(r, src_lb, src_ub, dst_lb, dst_ub)
+
+        vec = (y - x) / jnp.linalg.norm(y - x)
+        y_prime = x + new_r * vec
+
+        xy_prime = jnp.hstack([x, y_prime])
+        return xy_prime
+
+    def map_and_logdetjac(x, y, param):
+        xy = jnp.hstack([x, y])
+        xy_prime = f(xy, param)
+        y_prime = xy_prime[dim:]
+
+        jac = jacobian(f)(xy, param)
+        sign, logdet = jnp.linalg.slogdet(jac)
+
+        # negative det would be unexpected --> nan-poison if detected
+        logdetjac = jnp.where(sign == 1, logdet, jnp.nan)
+
+        return y_prime, logdetjac
+
+    x_b_mapped, logdetjac = map_and_logdetjac(x[a], x[b], param)
+    x_updated = x.at[b].set(x_b_mapped)
+
+    return x_updated, logdetjac
+
+
+apply_conf_map_to_traj = jit(vmap(conf_map, in_axes=(0, None, None)))
+
+
+def apply_conf_maps_to_traj(xs, bond_idxs, params) -> Tuple[Array, Array]:
+    """Apply maps to several bonds in a trajectory of conformers xs, return (updated xs, logdetjacs)"""
+    xs_traj, logdetjac_increments_traj = [jnp.array(xs)], [np.zeros(len(xs))]
+
+    for (bond, param) in zip(bond_idxs, params):  # TODO: jax.lax for-loop?
+        xs_updated, logdetjac_increments = apply_conf_map_to_traj(xs_traj[-1], bond, param)
+
+        xs_traj.append(xs_updated)
+        logdetjac_increments_traj.append(logdetjac_increments)
+
+    return xs_traj[-1], np.sum(logdetjac_increments_traj, axis=0)
+
+
+# utilities for getting terminal bonds
+def get_degrees(bond_idxs) -> Array:
+    g = nx.Graph()
+    g.add_edges_from(bond_idxs)
+    degrees = np.array([g.degree(i) for i in range(g.number_of_nodes())])
+    return degrees
+
+
+def get_terminal_bonds(bond_idxs) -> List[Tuple]:
+    """Get bonded pairs that involve a terminal atom"""
+
+    degrees = get_degrees(bond_idxs)
+    sort_by_descending_degree = lambda bond: tuple(sorted(bond, key=lambda i: degrees[i], reverse=True))
+    bonds = [sort_by_descending_degree(bond) for bond in bond_idxs]
+
+    is_terminal = lambda bond: degrees[bond[1]] == 1
+
+    terminal_bonds = sorted(filter(is_terminal, bonds))
+    return terminal_bonds
+
+
+class TerminalMappableState:
+    def __init__(self, terminal_bond_idxs, ks, eq_lengths, temperature=DEFAULT_TEMP, sigma_thresh=20):
```

Comment:
nit: Define DEFAULT_SIGMA_THRESHOLD=20 somewhere, since it appears repeatedly

Reply:
Addressed in https://github.com/proteneer/timemachine/pull/801/commits/d99f460f5eab2f0afb218e4cce8f26a75034d03d